overlord/hookstate/hookmgr.go

```diff
@@ -76,17 +77,45 @@ type HookSetup struct {
 func Manager(s *state.State) (*HookManager, error) {
 	runner := state.NewTaskRunner(s)
 	manager := &HookManager{
-		state:      s,
-		runner:     runner,
-		repository: newRepository(),
-		contexts:   make(map[string]*Context),
+		state:        s,
+		runner:       runner,
+		repository:   newRepository(),
+		contexts:     make(map[string]*Context),
+		snapContexts: newSnapContexts(s),
```

Comment:
We actually don't want these saved in memory persistently. Every time a snapctl call comes over, we need a new context which should live during that call only (thus ephemeral), otherwise two concurrent snapctl calls will corrupt each other's state. This is similar to how hooks work: two hooks, two contexts, even though the snap and the hook may be exactly the same.